File: String/LC290-WordPattern.cpp
class Solution {
public:
    bool wordPattern(string pattern, string str) {
        unordered_map<char, string> map;
        unordered_map<string, char> rmap;
        stringstream ss(str);
        string s;
        for (char c : pattern) {
            if (!(ss >> s) || (map.count(c) == 1 && map[c] != s) ||
               (rmap.count(s) == 1 && rmap[s] != c)) return false;
            map[c] = s;
            rmap[s] = c;
        }
        return (ss >> s) ? false : true;
    }
};

File: HashTable/LC290-WordPattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        # mapping: pattern to s
        # rmapping: s to pattern
        mapping = dict()
        rmapping = dict()
        # Index for cropping out the corresponding str in s
        start = 0
        end = 0
        # Go through all elements in the pattern
        # and find out the corresponding part same to the mapping or not
        for i in pattern:
            # move end to the space
            while end < len(s) and s[end] != " ":
                end += 1
            # extract the substring
            substr = s[start:end]
            # check the pattern:
            # if pattern in dict, check the word in s is same or not
            # if pattern not in dict, make it
            if i not in mapping:
                mapping[i] = substr
            else:
                if mapping[i] != substr:
                    return False
            # check the reverse mapping:
            if substr not in rmapping:
                rmapping[substr] = i
            else:
                if rmapping[substr] != i:
                    return False
            # Move the start and end to the beginning of next word
            start = end + 1
            end = start
        # I have remaining elements
        if end != len(s) + 1: return False
        return True
                
